### PR TITLE
757 s3 head

### PR DESCRIPTION
**Pull request by alaric-dotmesh**

### CODE REVIEW

File: cmd/dotmesh-server/s3_handlers.go [hunk 1]
```diff
@@ -205,6 +208,71 @@ func (s *S3Handler) mountFilesystemSnapshot(filesystemId string, snapshotId stri
...
+		http.Error(resp, fmt.Sprintf("please try again later, state was %s", state), http.StatusServiceUnavailable)
+		return
+	}
+
+	// we must first mount the given snapshot before we try to read a file within it
+	// if snapshotId is not given then the latest snapshot id will be used
+	e := s.mountFilesystemSnapshot(filesystemId, snapshotId)
+	// the snapshot has been mounted - pass the SnapshotMountPath via the
+	// OutputFile to the fileOutputIO channel to get handled
+	if e.Name == "mounted" {
```
Comment (Godley) [inline cmd/dotmesh-server/s3_handlers.go]: I know this is likely taken from the other func which doesn't do early return, but since go's style is to early return maybe change this to
```
if e.Name != "mounted" {
// chunk of code from the else block below
}
// chunk of code from inside the if statement
```

File: pkg/fsm/fsm_active_file_io.go [hunk 2]
```diff
@@ -95,6 +95,44 @@ func (f *FsMachine) saveFile(file *types.InputFile) StateFn {
...
+
+	// create the default paths
+	sourcePath := fmt.Sprintf("%s/%s/%s", file.SnapshotMountPath, "__default__", file.Filename)
+
+	l := log.WithFields(log.Fields{
+		"filename":   file.Filename,
+		"sourcePath": sourcePath,
+	})
+
+	fi, err := os.Stat(sourcePath)
```
Comment (Godley) [inline pkg/fsm/fsm_active_file_io.go]: is this still ok if it turns out to be a folder?

File: pkg/fsm/fsm_active_file_io.go [hunk 3]
```diff
@@ -95,6 +95,44 @@ func (f *FsMachine) saveFile(file *types.InputFile) StateFn {
 	return activeState
 }
 
+func (f *FsMachine) statFile(file *types.StatFile) StateFn {
+
+	// create the default paths
+	sourcePath := fmt.Sprintf("%s/%s/%s", file.SnapshotMountPath, "__default__", file.Filename)
```
Comment (rusenask) [inline pkg/fsm/fsm_active_file_io.go]: there's a method to do this without templates: `filepath.Join()`

File: tests/s3api_test.go [hunk 4]
```diff
@@ -373,30 +412,39 @@ func OSReadDir(root string) ([]string, error) {
 }
 
 func call(method string, path string, node citools.Node, body io.Reader) (respBody string, statusCode int, err error) {
```
Comment (Godley) [inline tests/s3api_test.go]: 1. why is this necessary? Doesn't that indicate a flake we should investigate (retrying)
2. Can you rename this function to indicate that it's retried?

File: cmd/dotmesh-server/s3_handlers.go (hunk 1 above)
Comment (rusenask) [inline cmd/dotmesh-server/s3_handlers.go]: might be nice to keep normal code path at minimal indentation (see: https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow).

If not mounted then just return error and continue without else?

File: cmd/dotmesh-server/s3_handlers.go [hunk 5]
```diff
@@ -205,6 +208,71 @@ func (s *S3Handler) mountFilesystemSnapshot(filesystemId string, snapshotId stri
...
+			http.Error(resp, "read failed, could not retrieve actual error", 500)
+		case types.EventNameFileNotFound:
+			err := result.Error()
+			if err != nil {
+				http.Error(resp, err.Error(), 404)
+				return
+			}
+			http.Error(resp, "file not found, could not retrieve actual error", 404)
+		default:
+			if (*result.Args)["mode"].(os.FileMode).IsRegular() {
```
Comment (rusenask) [inline cmd/dotmesh-server/s3_handlers.go]: would be good to check if it's set before converting to a type and calling a function or else it might panic :)

File: pkg/fsm/fsm_active_file_io.go (hunk 2 above)
Comment (alaric-dotmesh) [reply pkg/fsm/fsm_active_file_io.go]: Yeah, you can stat directories - the mode field of the FileInfo tells you if it's a directory, regular file, symlink, or other weirder thing (socket, device file, etc)

File: tests/s3api_test.go (hunk 4 above)
Comment (alaric-dotmesh) [reply tests/s3api_test.go]: There's a code path in the S3 handlers that return that code with a message to please try again, as the FSM hasn't returned to the active state yet, so I made sure the client tried again in that case. It cropped up in the test suite, which of course fires requests at the FSMs without any delays, so I reckon it could turn up in real life too!

And sure, I'll make it clear in the function name!

File: pkg/fsm/fsm_active_file_io.go (hunk 3 above)
Comment (alaric-dotmesh) [reply pkg/fsm/fsm_active_file_io.go]: There is indeed! I shameless cribbed this bit from the other functions in this file without improving any of them, but I can fix that now...
